com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ServiceRemoval.java

```diff
@@ -0,0 +1,138 @@
+package com.sap.cloud.lm.sl.cf.process.util;
+
+import java.text.MessageFormat;
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.cloudfoundry.client.lib.CloudControllerClient;
+import org.cloudfoundry.client.lib.CloudControllerException;
+import org.cloudfoundry.client.lib.CloudException;
+import org.cloudfoundry.client.lib.CloudOperationException;
+import org.cloudfoundry.client.lib.CloudServiceBrokerException;
+import org.cloudfoundry.client.lib.domain.CloudApplication;
+import org.cloudfoundry.client.lib.domain.CloudService;
+import org.cloudfoundry.client.lib.domain.CloudServiceBinding;
+import org.cloudfoundry.client.lib.domain.CloudServiceInstance;
+import org.cloudfoundry.client.lib.domain.CloudServiceKey;
+import org.springframework.http.HttpStatus;
+
+import com.sap.cloud.lm.sl.cf.core.model.ServiceInstanceWithServiceKeys;
+import com.sap.cloud.lm.sl.cf.core.security.serialization.SecureSerializationFacade;
+import com.sap.cloud.lm.sl.cf.core.util.ApplicationConfiguration;
+import com.sap.cloud.lm.sl.cf.process.Constants;
+import com.sap.cloud.lm.sl.cf.process.Messages;
+import com.sap.cloud.lm.sl.cf.process.steps.ExecutionWrapper;
+import com.sap.cloud.lm.sl.cf.process.steps.StepsUtil;
+import com.sap.cloud.lm.sl.cf.process.util.ExceptionMessageTailMapper.CloudComponents;
+import com.sap.cloud.lm.sl.common.SLException;
+
+@Named
+public class ServiceRemoval {
+
+    private final SecureSerializationFacade secureSerializer = new SecureSerializationFacade();
+
+    private ApplicationConfiguration configuration;
+
+    @Inject
+    public ServiceRemoval(ApplicationConfiguration configuration) {
+        this.configuration = configuration;
+    }
+
+    public void deleteService(ExecutionWrapper execution, ServiceInstanceWithServiceKeys service) {
+        CloudControllerClient client = execution.getControllerClient();
+
+        try {
+            prepareServicesToDelete(client, execution.getStepLogger(), service);
+            deleteService(client, execution.getStepLogger(), service);
+        } catch (CloudException e) {
+            processException(execution, e, client.getServiceInstance(service.getServiceInstance()
+                                                                            .getName()));
+        }
+
+    }
+
+    private void prepareServicesToDelete(CloudControllerClient client, StepLogger stepLogger, ServiceInstanceWithServiceKeys service) {
+        unbindService(client, stepLogger, service.getServiceInstance());
+        deleteServiceKeys(client, stepLogger, service.getServiceKeys());
+
+    }
+
+    private void unbindService(CloudControllerClient client, StepLogger stepLogger, CloudServiceInstance serviceInstance) {
+        List<CloudServiceBinding> bindings = serviceInstance.getBindings();
+        if (bindings.isEmpty()) {
```

Comment:
Why are those checks needed? You check for empty associations before you call deleteService. If there are no empty associations - would this ever be reached at all?